mesh/src/random_field/OffLatticeRandomFieldGenerator.cpp

```diff
@@ -0,0 +1,151 @@
+IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE
+GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
+HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT
+OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+*/
+
+#include "OffLatticeRandomFieldGenerator.hpp"
+
+#include <algorithm>
+#include <iomanip>
+#include <fstream>
+#include <memory>
+#include <numeric>
+
+#include "Exception.hpp"
+#include "RandomNumberGenerator.hpp"
+
+template <unsigned SPACE_DIM>
+OffLatticeRandomFieldGenerator<SPACE_DIM>::OffLatticeRandomFieldGenerator(
+    std::array<double, SPACE_DIM> lowerCorner,
+    std::array<double, SPACE_DIM> upperCorner,
+    std::array<bool, SPACE_DIM> periodicity,
+    double lengthScale,
+    double boxWidth)
+    : mLowerCorner(lowerCorner),
+      mUpperCorner(upperCorner),
+      mPeriodicity(periodicity),
+      mLengthScale(lengthScale),
+      mpBoxCollection(nullptr)
+{
+    // Reset the box width if the default value is being used
+    if (boxWidth == DOUBLE_UNSET)
+    {
+        const double max_box_width = 4.0 * lengthScale; 
+        boxWidth = std::min(max_box_width, upperCorner[0] - lowerCorner[0]);
+    }
+
+    // Set up the box collection
+    c_vector<double, 2 * SPACE_DIM> domain_size;
+    for (unsigned dim = 0; dim < SPACE_DIM; ++dim)
+    {
+        domain_size[2 * dim] = lowerCorner[dim];
+        domain_size[2 * dim + 1] = upperCorner[dim];
+    }
+
+    ///\todo: Obsolete box collection should take a std::array<bool, DIM>
+    bool periodic_x = periodicity[0];
+    bool periodic_y = SPACE_DIM > 1 ? periodicity[1] : false;
+    bool periodic_z = SPACE_DIM > 2 ? periodicity[2] : false;
+
+    mpBoxCollection = std::make_unique<ObsoleteBoxCollection<SPACE_DIM>>(
+        boxWidth,
+        domain_size,
+        periodic_x,
+        periodic_y,
+        periodic_z
+    );
+
+    mpBoxCollection->SetupLocalBoxesHalfOnly();    
+    mOpenSimplex = OpenSimplex2S();
+}
+
+template <unsigned SPACE_DIM>
+void OffLatticeRandomFieldGenerator<SPACE_DIM>::SetRandomSeed(
+    const unsigned seed)
+{
+    mOpenSimplex = OpenSimplex2S(seed);
+}
+
+template <unsigned SPACE_DIM>
+std::vector<double> OffLatticeRandomFieldGenerator<SPACE_DIM>::SampleRandomField(
+    const std::vector<Node<SPACE_DIM>*>& rNodes)
+{
+    return this->SampleRandomFieldAtTime(rNodes, 0.0);
+}
+
+template <unsigned SPACE_DIM>
+std::vector<double> OffLatticeRandomFieldGenerator<SPACE_DIM>::SampleRandomFieldAtTime(
+    const std::vector<Node<SPACE_DIM>*>& rNodes,
+    const double time)
+{
+    auto reshape = [](const double val)
```

Comment:
Move this out into a separate utility header etc. as a named function with extensive description in comments.